core/templates/services/user-backend-api.service.ts

```diff
@@ -0,0 +1,157 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for user data.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { WindowRef } from 'services/contextual/window-ref.service';
+
+import { UrlInterpolationService } from
+  'domain/utilities/url-interpolation.service';
+import { UrlService } from 'services/contextual/url.service.ts';
+import { UserInfoObjectFactory, UserInfo, IUserInfoBackendDict } from
+  'domain/user/UserInfoObjectFactory.ts';
+import { AppConstants } from 'app.constants';
+import { List } from 'lodash';
+
+interface subscriptionSummary {
+  'creator_picture_data_url': string;
+  'creator_username': string;
+  'creator_impact': number;
+}
+interface IPreferencesBackendDict {
+  'preferred_language_codes': string;
+  'preferred_site_language_code': string;
+  'preferred_audio_language_code': string;
+  'profile_picture_data_url': string;
+  'default_dashboard': string;
+  'user_bio': string;
+  'subject_interests': string;
+  'can_receive_email_updates': boolean;
+  'can_receive_editor_role_email': boolean;
+  'can_receive_feedback_message_email': boolean;
+  'can_receive_subscription_email': boolean;
+  'subscription_list': List<subscriptionSummary>;
+}
+interface IUrlBackendDict {
+  'login_url': string;
+}
+interface IUserCommunityRightsDataBackendDict {
+  'can_review_translation_for_language_codes': boolean;
+  'can_review_voiceover_for_language_codes': boolean;
+  'can_review_questions': boolean;
+}
+
+
+@Injectable({
+  providedIn: 'root'
+})
+export class UserBackendApiService {
+  constructor(
+    private http: HttpClient,
+    private windowRef: WindowRef,
+    private urlInterpolationService: UrlInterpolationService,
+    private urlService: UrlService,
+    private userInfoObjectFactory: UserInfoObjectFactory) {}
+
+    private PREFERENCES_DATA_URL = '/preferenceshandler/data';
+    private USER_COMMUNITY_RIGHTS_DATA_URL = '/usercommunityrightsdatahandler';
+
+    private userInfo = null;
+    private userCommunityRightsInfo = null;
+
+    getUserInfoAsync(): Promise<UserInfo> {
+      if (this.urlService.getPathname() === '/signup') {
+        return Promise.resolve(this.userInfoObjectFactory.createDefault());
+      }
+      if (this.userInfo) {
+        return Promise.resolve(this.userInfo);
+      }
+      return this.http.get<IUserInfoBackendDict>(
+        '/userinfohandler').toPromise().then(
+        (backendDict) => {
+          if (backendDict.user_is_logged_in) {
+            this.userInfo =
+              this.userInfoObjectFactory.createFromBackendDict(
+                backendDict);
+            return Promise.resolve(this.userInfo);
+          } else {
+            return Promise.resolve(
+              this.userInfoObjectFactory.createDefault());
+          }
+        });
+    }
```

Comment:
You need to create a new promise. Unlike `$q` where `$q.resolve` works, here you have to make a new promise and resolve it. Please make this change here and everywhere else in this file.
```suggestion
    getUserInfoAsync(): Promise<UserInfo> {
      return new Promise((resolve, reject) => {
        if (this.urlService.getPathname() === '/signup') {
          return resolve(this.userInfoObjectFactory.createDefault());
        }
        if (this.userInfo) {
          return resolve(this.userInfo);
        }
        return this.http.get<IUserInfoBackendDict>(
          '/userinfohandler').toPromise().then(
          (backendDict) => {
            if (backendDict.user_is_logged_in) {
              this.userInfo =
                this.userInfoObjectFactory.createFromBackendDict(
                  backendDict);
              return resolve(this.userInfo);
            } else {
              return resolve(
                this.userInfoObjectFactory.createDefault());
            }
          });
        });
    }
```